### PR TITLE
IDMP-659 - add package components, medical device and characteristics

### PR DESCRIPTION
**Pull request by tw-osthus**
Description: add package components, medical device and associated physical characteristics in idmp-mprd, with more general classes for materials in idmp-sub

### CODE REVIEW

**Comment by tw-osthus** [comment]
volume is used as physical characteristic class in mprd, and as quantity kind instance in ucum. @ElisaKendall how can be solve this conflict?

**Comment by ElisaKendall** [comment]
> volume is used as physical characteristic class in mprd, and as quantity kind instance in ucum. @ElisaKendall how can be solve this conflict?

My suggestion is that volume and other quantity kinds should be modeled as instances rather than as classes, following the approach that is taken in the OMG Quantities and Units ontology and related UCUM examples, rather than re-modeling volume as a class. I will take a look at the standard and see what is required, but to model every quantity kind, dimension, and unit as a class would produce a model that requires everyone using it to create instances of these classes, which is not efficient or sustainable.

File: ISO/ISO11615-MedicinalProducts.rdf [hunk 1]
```diff
@@ -3083,6 +3238,139 @@ g) therapeutic indication(s) as authorized for the medicinal product.</skos:note
...
+			</owl:Restriction>
+		</rdfs:subClassOf>
+		<rdfs:label>width characteristic</rdfs:label>
+		<skos:definition>1-D extent physical characteristic which is equal to the distance from one side of an object to another side which is opposite</skos:definition>
+		<cmns-av:adaptedFrom>ISO 11615:2017 Health informatics - Identification of medicinal products (IDMP) - Data elements and structures for the unique identification and exchange of regulated medicinal product information, clause 9.6.2.21.3</cmns-av:adaptedFrom>
+		<cmns-av:adaptedFrom rdf:datatype="&xsd;anyURI">http://purl.obolibrary.org/obo/PATO_000001595</cmns-av:adaptedFrom>
+		<cmns-av:synonym>width</cmns-av:synonym>
+	</owl:Class>
+	
+	<owl:Class rdf:about="&idmp-mprd;WithPhysicalCharacteristics">
```
Comment (mereolog) [inline ISO/ISO11615-MedicinalProducts.rdf:3289]: @tw-osthus can we have the name that starts with a noun, e.g., ThingWithPhysicalCharacteristics?

Comment (mereolog) [inline ISO/ISO11615-MedicinalProducts.rdf:3289]: Can we have some more demanding restriction, e.g., like  "isCharacterizedBy some Physical Characteristic"?

**Review by mereolog — COMMENTED** [review]
The way that physical characteristics are modelled assumes that things do not change (wrt these characteristics, e.g., wrt weight).

Perhaps we should be explicit about that or start modelling something like states (of things) or their temporal parts.

File: ISO/ISO11615-MedicinalProducts.rdf (hunk 1 above)
Comment (tw-osthus) [reply ISO/ISO11615-MedicinalProducts.rdf:3289]: physical characteristics in ISO-IDMP is entity with the properties that I have added with minQualifiedCardinality. It makes sense that something with physical characteristic has at least one physical characteristic.

Comment (tw-osthus) [reply ISO/ISO11615-MedicinalProducts.rdf]: done

**Review by mereolog — DISMISSED** [review]

File: ISO/ISO11615-MedicinalProducts.rdf (hunk 1 above)
Comment (ElisaKendall) [reply ISO/ISO11615-MedicinalProducts.rdf]: @tw-osthus @mereolog I think there is a deeper issue here - the distinction between something being a specification for something and being a thing in the real world is starting to catch up with us. Something that is a specification is describing the thing to be manufactured, whereas the actual thing in the world has physical characteristics. My recommendation is that we make sure we keep these separate, and this approach is attempting to skirt that in ways that will catch up with us. If something has physical characteristics, it is matter. The restriction that says something has physical characteristics should be on matter. These other characteristics of some specification of the thing to be built should be specified as characteristics (aspects) but not as physical characteristics. We used the notion of a defining characteristic for specifications of substances, doesn't that work here to distinguish the two kinds of characteristics?

Comment (tw-osthus) [reply ISO/ISO11615-MedicinalProducts.rdf]: In this context, the physical characteristics are specifications or descriptions of physical characteristics of matter. Amount is also a physical characteristic and exists already a while in the ontology.
Note that the definition of physical characteristic is
> characterization of matter (or a specification thereof) ...
allows this use case.

Comment (tw-osthus) [reply ISO/ISO11615-MedicinalProducts.rdf]: I do not like  "ThingsWithPhysicalCharacteristics" as well, but it is a defined class that represents the "physical characteristics" entity in the ISO IDMP standard. The properties collected in it are shared between 3 different entities, so the class simplifies the ontology and ensures the same structure as in the standard. It is not useless. I have added a note, that it should not be used outside the ontology. Note that the properties are optional, except the one restriction, that adds that a thing with physical characteristic must have a physical characteristic, which is not part of the standard and only in the ontology. 

We can go without this technical class and copy the restrictions into the 3 classes directly.

If we want to forbid defined classes in the ontology, that are only used to introduce aspects to other classes, e.g. adding certain set of restrictions, without being universals, then that can make ontology design more inefficient and less robust. I  compare these defined classes to aspects in aspect oriented programming. They are shared between classes (in object oriented programming) and are technically classes themselves, but are not domain entities.

Comment (tw-osthus) [reply ISO/ISO11615-MedicinalProducts.rdf]: Again physical characteristic **as defined**  in IDMP-O is not restricted to physical stuff, but can be used to specify physical stuff. In fact, the use of the ontology is only in this sense of specification. The use of physical characteristics such as amount on the physical matter is only applied on measuring, which will apply in studies or in the manufacturing or validation of products, but not in regulation affairs.

As specifications about physical stuff is what is the domain of IDMP, we made this compromise. Personally I'd have preferred that we keep the distinction, and in Allotrope we certainly kept them separate. However Allotrope is a different domain, its focus is on measuring. In IDMP we have recognized, that the strict distinction makes the ontology to complicated for the IDMP domain and so we have decided to relax the distinctions between specification and the real world. We cannot have the cake and eat it too.

Using "comprises" as relation between package items to express containment is far worse then this and we allowed and recommended it in order to simplify the use of the IDMP ontology. Package items are specifications about physical containers. Containers contain other things, not their specifications. Specifications may contain sub-specifications and propositions.

Comment (ElisaKendall) [reply ISO/ISO11615-MedicinalProducts.rdf]: > In this context, the physical characteristics are specifications or descriptions of physical characteristics of matter. Amount is also a physical characteristic and exists already a while in the ontology. Note that the definition of physical characteristic is
> 
> > characterization of matter (or a specification thereof) ...
> 
> allows this use case.

Yes, I get that and recognized it when I made the comment, but I am still concerned a bit that users may struggle with this. I understand the separation of concerns and why it was done in Allotrope, but hopefully we can be clear enough in IDMP that people are not confused.

Comment (ElisaKendall) [reply ISO/ISO11615-MedicinalProducts.rdf]: > I do not like "ThingsWithPhysicalCharacteristics" as well, but it is a defined class that represents the "physical characteristics" entity in the ISO IDMP standard. The properties collected in it are shared between 3 different entities, so the class simplifies the ontology and ensures the same structure as in the standard. It is not useless. I have added a note, that it should not be used outside the ontology. Note that the properties are optional, except the one restriction, that adds that a thing with physical characteristic must have a physical characteristic, which is not part of the standard and only in the ontology.
> 
> We can go without this technical class and copy the restrictions into the 3 classes directly.
> 
> If we want to forbid defined classes in the ontology, that are only used to introduce aspects to other classes, e.g. adding certain set of restrictions, without being universals, then that can make ontology design more inefficient and less robust. I compare these defined classes to aspects in aspect oriented programming. They are shared between classes (in object oriented programming) and are technically classes themselves, but are not domain entities.

I'm not ready to rule out defined classes where appropriate. But this class stood out to me due to its name and the bag of characteristics in it. I think it is essentially a collection of characteristics, which is a "device" used to simplify the relational model, I guess is the point I was trying to make. Adding the one restriction that says specifications of products (medicinal and pharmaceutical?) have some physical characteristic, and then keeping the classes of characteristics as subclasses of physical characteristic seems like the right model to me. Then users can add whichever ones apply to a given product.

**Review by ElisaKendall — APPROVED** [review]
This is fine with the most recent changes applied.

**Review by mereolog — APPROVED** [review]

**Pull request merged**